File: src/exact.js
// --- BigInt helpers -------------------------------------------------------

const TEN = 10n;

function absBigInt(value) {
  return value < 0n ? -value : value;
}

function gcd(a, b) {
  a = absBigInt(a);
  b = absBigInt(b);
  while (b) [a, b] = [b, a % b];
  return a || 1n;
}

export function floorDiv(a, b) {
  if (b === 0n) throw new RangeError("Division by zero");
  let quotient = a / b;
  const remainder = a % b;
  if (remainder !== 0n && (remainder < 0n) !== (b < 0n)) quotient -= 1n;
  return quotient;
}

export function floorMod(a, b) {
  return a - floorDiv(a, b) * b;
}

// --- Rational: arbitrary-precision exact fractions -------------------------

export class Rational {
  constructor(numerator, denominator = 1n) {
    let n = typeof numerator === "bigint" ? numerator : BigInt(numerator);
    let d = typeof denominator === "bigint" ? denominator : BigInt(denominator);
    if (d === 0n) throw new RangeError("Division by zero");
    if (d < 0n) {
      n = -n;
      d = -d;
    }
    const divisor = gcd(n, d);
    this.n = n / divisor;
    this.d = d / divisor;
    Object.freeze(this);
  }

  static parse(value) {
    if (value instanceof Rational) return value;
    if (typeof value === "bigint") return new Rational(value);
    if (typeof value === "number") {
      if (!Number.isFinite(value)) throw new TypeError("A rational must be finite");
      return Rational.parse(String(value));
    }
    const text = String(value).trim();
    const fraction = /^([+-]?\d+)\/(\d+)$/.exec(text);
    if (fraction) return new Rational(BigInt(fraction[1]), BigInt(fraction[2]));
    const decimal = /^([+-]?)(\d*)(?:\.(\d*))?(?:[eE]([+-]?\d+))?$/.exec(text);
    if (!decimal || (!decimal[2] && !decimal[3])) throw new TypeError(`Invalid exact number: ${text}`);
    const sign = decimal[1] === "-" ? -1n : 1n;
    const whole = decimal[2] || "0";
    const part = decimal[3] || "";
    const exponent = Number(decimal[4] || 0) - part.length;
    const digits = BigInt((whole + part).replace(/^0+(?=\d)/, "") || "0") * sign;
    return exponent >= 0
      ? new Rational(digits * TEN ** BigInt(exponent))
      : new Rational(digits, TEN ** BigInt(-exponent));
  }

  add(other) {
    const value = Rational.parse(other);
    return new Rational(this.n * value.d + value.n * this.d, this.d * value.d);
  }

  sub(other) {
    const value = Rational.parse(other);
    return new Rational(this.n * value.d - value.n * this.d, this.d * value.d);
  }

  mul(other) {
    const value = Rational.parse(other);
    return new Rational(this.n * value.n, this.d * value.d);
  }

  div(other) {
    const value = Rational.parse(other);
    return new Rational(this.n * value.d, this.d * value.n);
  }

  neg() {
    return new Rational(-this.n, this.d);
  }

  abs() {
    return new Rational(absBigInt(this.n), this.d);
  }

  pow(exponent) {
    const power = typeof exponent === "bigint" ? exponent : BigInt(exponent);
    if (power < 0n) return new Rational(this.d ** -power, this.n ** -power);
    return new Rational(this.n ** power, this.d ** power);
  }

  floor() {
    return floorDiv(this.n, this.d);
  }

  ceil() {
    return -floorDiv(-this.n, this.d);
  }

  mod(other) {
    const value = Rational.parse(other);
    return this.sub(value.mul(this.div(value).floor()));
  }

  compare(other) {
    const value = Rational.parse(other);
    const delta = this.n * value.d - value.n * this.d;
    return delta < 0n ? -1 : delta > 0n ? 1 : 0;
  }

  isZero() {
    return this.n === 0n;
  }

  toNumber() {
    return Number(this.n) / Number(this.d);
  }

  toDecimal(places = 18) {
    const negative = this.n < 0n;
    const numerator = absBigInt(this.n);
    const whole = numerator / this.d;
    let remainder = numerator % this.d;
    if (!remainder) return `${negative ? "-" : ""}${whole}`;
    let digits = "";
    for (let index = 0; index < places && remainder; index += 1) {
      remainder *= 10n;
      digits += remainder / this.d;
      remainder %= this.d;
    }
    return `${negative ? "-" : ""}${whole}.${digits.replace(/0+$/, "") || "0"}`;
  }

  toJSON() {
    return this.d === 1n ? this.n.toString() : `${this.n}/${this.d}`;
  }

  toString() {
    return this.toJSON();
  }
}

export const ZERO = new Rational(0n);
export const ONE = new Rational(1n);

// A fixed, versioned π constant makes range reduction independent of host Math.
export const PI = Rational.parse(
  "3.141592653589793238462643383279502884197169399375105820974944592307816406286"
);
export const TAU = PI.mul(2);

// --- Transcendental functions (fixed-point series, no host Math) ----------

function roundRational(value, decimalPlaces) {
  const rational = Rational.parse(value);
  const scale = TEN ** BigInt(decimalPlaces);
  const scaled = rational.mul(scale);
  const half = new Rational(1n, 2n);
  const rounded = scaled.n < 0n
    ? scaled.sub(half).ceil()
    : scaled.add(half).floor();
  return new Rational(rounded, scale);
}

export function sinExact(value, decimalPlaces = 30) {
  const workingPlaces = decimalPlaces + 8;
  let x = roundRational(Rational.parse(value).mod(TAU), workingPlaces);
  if (x.compare(PI) > 0) x = x.sub(TAU);
  if (x.compare(PI.neg()) < 0) x = x.add(TAU);

  let term = x;
  let sum = x;
  const limit = new Rational(1n, TEN ** BigInt(workingPlaces));
  for (let index = 1n; index < 256n; index += 1n) {
    const divisor = (2n * index) * (2n * index + 1n);
    term = roundRational(term.mul(x).mul(x).div(divisor).neg(), workingPlaces);
    sum = roundRational(sum.add(term), workingPlaces);
    if (term.abs().compare(limit) < 0) break;
  }
  return roundRational(sum, decimalPlaces);
}

export function cosExact(value, decimalPlaces = 30) {
  return sinExact(Rational.parse(value).add(PI.div(2)), decimalPlaces);
}

export function sqrtExact(value, decimalPlaces = 30) {
  const input = Rational.parse(value);
  if (input.compare(0) < 0) throw new RangeError("Square root of a negative value");
  if (input.isZero()) return ZERO;
  const workingPlaces = decimalPlaces + 8;
  const exponent = BigInt(input.n.toString(2).length - input.d.toString(2).length);
  const half = exponent >> 1n;
  let guess = half >= 0n ? new Rational(2n ** half) : new Rational(1n, 2n ** -half);
  const limit = new Rational(1n, TEN ** BigInt(decimalPlaces + 6));
  for (let index = 0; index < 128; index += 1) {
    const next = roundRational(guess.add(input.div(guess)).div(2), workingPlaces);
    if (next.isZero()) return ZERO;
    if (next.sub(guess).abs().compare(limit) < 0) {
      guess = next;
      break;
    }
    guess = next;
  }
  return roundRational(guess, decimalPlaces);
}

// --- Gregorian calendar conversion -----------------------------------------
// Proleptic Gregorian conversion using arbitrary-size integers.
// Day zero is 1970-01-01, but callers never need to expose that implementation origin.
export function daysFromCivil(yearValue, monthValue, dayValue) {
  let year = BigInt(yearValue);
  const month = BigInt(monthValue);
  const day = BigInt(dayValue);
  year -= month <= 2n ? 1n : 0n;
  const era = floorDiv(year, 400n);
  const yearOfEra = year - era * 400n;
  const shiftedMonth = month + (month > 2n ? -3n : 9n);
  const dayOfYear = floorDiv(153n * shiftedMonth + 2n, 5n) + day - 1n;
  const dayOfEra = yearOfEra * 365n + floorDiv(yearOfEra, 4n)
    - floorDiv(yearOfEra, 100n) + dayOfYear;
  return era * 146097n + dayOfEra - 719468n;
}

export function civilFromDays(dayValue) {
  const z = BigInt(dayValue) + 719468n;
  const era = floorDiv(z, 146097n);
  const dayOfEra = z - era * 146097n;
  const yearOfEra = floorDiv(
    dayOfEra - floorDiv(dayOfEra, 1460n) + floorDiv(dayOfEra, 36524n)
      - floorDiv(dayOfEra, 146096n),
    365n
  );
  let year = yearOfEra + era * 400n;
  const dayOfYear = dayOfEra - (
    365n * yearOfEra + floorDiv(yearOfEra, 4n) - floorDiv(yearOfEra, 100n)
  );
  const monthPrime = floorDiv(5n * dayOfYear + 2n, 153n);
  const day = dayOfYear - floorDiv(153n * monthPrime + 2n, 5n) + 1n;
  const month = monthPrime + (monthPrime < 10n ? 3n : -9n);
  year += month <= 2n ? 1n : 0n;
  return { year, month, day };
}

export function isLeapYear(yearValue) {
  const year = BigInt(yearValue);
  return floorMod(year, 4n) === 0n
    && (floorMod(year, 100n) !== 0n || floorMod(year, 400n) === 0n);
}

export function daysInMonth(yearValue, monthValue) {
  const month = Number(monthValue);
  if (month === 2) return isLeapYear(yearValue) ? 29 : 28;
  return [4, 6, 9, 11].includes(month) ? 30 : 31;
}

// The single sanctioned boundary where host wall-clock time enters domain
// math: this is the one place `new Date()` is read to mean "now" anywhere in
// ChronoLog. It converts that one read into an exact local-civil day ordinal
// (day + hour/minute/second fraction) and never touches `Date` again. AGENTS.md's
// "no Date arithmetic in domain code" rule only holds if every caller shares
// this conversion instead of rebuilding it inline — three call sites once
// duplicated this exact arithmetic (two at minute precision, one at second
// precision); melting them here is what makes the rule enforceable rather
// than aspirational. Do not "fix" this back into per-call-site Date reads.
// Second precision is the canonical one: it is the finer of the two
// precisions that existed, and nothing was found to depend on minute
// truncation (see roster.js/session.js callers). A caller that genuinely
// needs a coarser ordinal should truncate this result explicitly at its own
// call site, not fork a second helper.
// `clock` is an optional injected `Date` (or Date-constructible value) so
// this stays testable without mocking the global `Date`.
// The 24/1440/86400 below are NOT the document's coordinate law and must never
// be melted into it: they are the host operating system's own civil units, the
// only units a `Date` can report. This converts a host reading into a day
// ordinal on the shared days axis; how that ordinal then reads as hours is the
// governing frame's law to decide (src/coordinate-law.js). Same boundary rule as
// ICS: the outside world speaks standard civil time, and conversion happens once
// on the way in.
export function nowDays(clock = new Date()) {
  const date = clock instanceof Date ? clock : new Date(clock);
  return new Rational(daysFromCivil(
    BigInt(date.getFullYear()),
    BigInt(date.getMonth() + 1),
    BigInt(date.getDate())
  ))
    .add(Rational.parse(date.getHours()).div(24))
    .add(Rational.parse(date.getMinutes()).div(1440))
    .add(Rational.parse(date.getSeconds()).div(86400));
}

// --- Nested-coordinate helpers ----------------------------------------------

export function coordinate(levels) {
  return {
    levels: levels
      .filter((entry) => entry.value !== undefined && entry.value !== null)
      .map((entry) => ({ level: String(entry.level), value: String(entry.value) }))
  };
}

export function levelValue(value, name, fallback = "0") {
  const entry = value?.levels?.find((part) => part.level === name);
  return entry ? entry.value : fallback;
}

// Duration magnitudes, nested-coordinate conversion, and the unit ladder that
// used to live here now belong to src/coordinate-law.js: every one of them is a
// question about a FRAME'S DECLARED LAW ("how long is an hour here?"), not a
// question about exact arithmetic, and answering them from a fixed table here is
// what made a frame's levels/radix/transition ladder dead metadata. What remains
// in this module is the arithmetic itself plus the proleptic Gregorian whole-day
// kernel (`daysFromCivil`/`civilFromDays`/`daysInMonth`/`isLeapYear`), which the
// registered Gregorian family calls and nothing else may.
export function formatCivil(value, includeTime = false) {
  const year = levelValue(value, "year", "1970");
  const month = levelValue(value, "month", "1").padStart(2, "0");
  const day = levelValue(value, "day", "1").padStart(2, "0");
  if (!includeTime) return `${year}-${month}-${day}`;
  const hour = levelValue(value, "hour", "0").padStart(2, "0");
  const minute = levelValue(value, "minute", "0").padStart(2, "0");
  const second = levelValue(value, "second", "0").padStart(2, "0");
  return `${year}-${month}-${day} ${hour}:${minute}:${second}`;
}

File: app/lib/core/exact.dart
// Pure Dart. Nothing under lib/core/ may import Flutter or a third-party
// package. This file is the numeric substrate every document value rests on,
// so no host floating point may enter any operation that decides a value.

// Hoisted: the calendar kernel runs inside occurrence loops at overscale, and
// `BigInt.from` per call would allocate on every iteration.
final BigInt _b4 = BigInt.from(4),
    _b10 = BigInt.from(10),
    _b100 = BigInt.from(100),
    _b400 = BigInt.from(400),
    _b146097 = BigInt.from(146097),
    _b719468 = BigInt.from(719468);

/// Floored division. Dart's `~/` truncates toward zero and its `%` is
/// Euclidean; the calendar kernel and every modular reduction here need
/// floored semantics, so neither built-in can be substituted.
BigInt floorDiv(BigInt a, BigInt b) {
  if (b == BigInt.zero) throw RangeError('Division by zero');
  final quotient = a ~/ b;
  final remainder = a.remainder(b);
  final short = remainder != BigInt.zero && remainder.isNegative != b.isNegative;
  return short ? quotient - BigInt.one : quotient;
}

/// Remainder under floored division: the sign follows `b`, never `a`.
BigInt floorMod(BigInt a, BigInt b) => a - floorDiv(a, b) * b;

/// An exact fraction over arbitrary-precision integers. Immutable, and always
/// stored reduced with a positive denominator, so one value has exactly one
/// representation and structural equality agrees with [compareTo].
class Rational implements Comparable<Rational> {
  final BigInt n;
  final BigInt d;

  Rational._(this.n, this.d);

  factory Rational(BigInt numerator, [BigInt? denominator]) {
    final d = denominator ?? BigInt.one;
    if (d == BigInt.zero) throw RangeError('Division by zero');
    // A negatively signed gcd reduces and moves the sign out of the
    // denominator in one division.
    final g = numerator.gcd(d) * (d.isNegative ? -BigInt.one : BigInt.one);
    return Rational._(numerator ~/ g, d ~/ g);
  }

  factory Rational.fromInt(int numerator, [int denominator = 1]) =>
      Rational(BigInt.from(numerator), BigInt.from(denominator));

  /// Accepts the serialized form (`3`, `-1/4000`) and decimal or scientific
  /// notation. No double is constructed on this path, so `1e-330` parses
  /// exactly where a host float would flush it to zero.
  factory Rational.parse(String text) {
    final source = text.trim();
    final ratio = _fraction.firstMatch(source);
    if (ratio != null) {
      return Rational(BigInt.parse(ratio[1]!), BigInt.parse(ratio[2]!));
    }
    final m = _decimal.firstMatch(source);
    if (m == null || ((m[2] ?? '').isEmpty && (m[3] ?? '').isEmpty)) {
      throw FormatException('Invalid exact number', text);
    }
    final part = m[3] ?? '';
    final sign = m[1] == '-' ? '-' : '';
    final digits = BigInt.parse('$sign${m[2] ?? ''}$part');
    final power = int.parse(m[4] ?? '0') - part.length;
    return power >= 0 ? Rational(digits * _b10.pow(power)) : Rational(digits, _b10.pow(-power));
  }

  static final RegExp _fraction = RegExp(r'^([+-]?\d+)/(\d+)$');
  static final RegExp _decimal = RegExp(r'^([+-]?)(\d*)(?:\.(\d*))?(?:[eE]([+-]?\d+))?$');
  static final RegExp _trailingZeros = RegExp(r'0+$');

  static final Rational zero = Rational(BigInt.zero);
  static final Rational one = Rational(BigInt.one);
  static final Rational _half = Rational(BigInt.one, BigInt.two);

  Rational operator +(Rational o) => Rational(n * o.d + o.n * d, d * o.d);

  Rational operator -(Rational o) => Rational(n * o.d - o.n * d, d * o.d);

  Rational operator *(Rational o) => Rational(n * o.n, d * o.d);

  Rational operator /(Rational o) => Rational(n * o.d, d * o.n);

  // Negation cannot break the invariant, so it skips reduction.
  Rational operator -() => Rational._(-n, d);

  /// Remainder under floored division, matching [floorMod].
  Rational operator %(Rational o) => this - o * Rational((this / o).floor());

  Rational abs() => n.isNegative ? -this : this;

  Rational pow(int exponent) => exponent < 0
      ? Rational(d.pow(-exponent), n.pow(-exponent))
      : Rational(n.pow(exponent), d.pow(exponent));

  BigInt floor() => floorDiv(n, d);

  BigInt ceil() => -floorDiv(-n, d);

  /// Half away from zero: the only rounding the ported substrate used.
  BigInt round() => n.isNegative ? (this - _half).ceil() : (this + _half).floor();

  @override
  int compareTo(Rational o) => (n * o.d).compareTo(o.n * d);

  bool operator <(Rational o) => compareTo(o) < 0;

  bool operator <=(Rational o) => compareTo(o) <= 0;

  bool operator >(Rational o) => compareTo(o) > 0;

  bool operator >=(Rational o) => compareTo(o) >= 0;

  @override
  bool operator ==(Object other) => other is Rational && n == other.n && d == other.d;

  @override
  int get hashCode => Object.hash(n, d);

  bool get isZero => n == BigInt.zero;

  bool get isNegative => n.isNegative;

  /// Host float, for pixels only: a result of this must never re-enter
  /// document math.
  double toDouble() => n / d;

  /// Truncates rather than rounds, and drops trailing zeros, so any value
  /// expressible in [places] digits survives a [Rational.parse] round trip.
  String toDecimal([int places = 18]) {
    final sign = n.isNegative ? '-' : '';
    final whole = n.abs() ~/ d;
    var rest = n.abs().remainder(d);
    if (rest == BigInt.zero) return '$sign$whole';
    final digits = StringBuffer();
    for (var i = 0; i < places && rest != BigInt.zero; i++) {
      rest *= _b10;
      digits.write(rest ~/ d);
      rest = rest.remainder(d);
    }
    final tail = digits.toString().replaceAll(_trailingZeros, '');
    return '$sign$whole.${tail.isEmpty ? '0' : tail}';
  }

  /// The persisted form. Documents written by the JavaScript implementation
  /// carry exactly these two shapes, so this string is a compatibility
  /// contract, not a formatting choice.
  String toJson() => d == BigInt.one ? '$n' : '$n/$d';

  @override
  String toString() => toJson();
}

typedef CivilDay = ({BigInt year, int month, int day});

/// Proleptic Gregorian whole-day kernel. Day zero is 1970-01-01; that origin
/// is an implementation detail no caller needs to restate. The 400-year era
/// arithmetic cannot be derived from per-level counts, which is why it lives
/// here and not in a frame's declared law. Only the year is unbounded: every
/// intermediate below is proven non-negative and inside a 400-year era, so
/// `int` division is exactly the floored division the formula requires.
BigInt daysFromCivil(BigInt year, int month, int day) {
  final shifted = month <= 2 ? year - BigInt.one : year;
  final era = floorDiv(shifted, _b400);
  final yearOfEra = (shifted - era * _b400).toInt();
  final monthPrime = month + (month > 2 ? -3 : 9);
  final dayOfYear = (153 * monthPrime + 2) ~/ 5 + day - 1;
  final dayOfEra = yearOfEra * 365 + yearOfEra ~/ 4 - yearOfEra ~/ 100 + dayOfYear;
  return era * _b146097 + BigInt.from(dayOfEra) - _b719468;
}

CivilDay civilFromDays(BigInt days) {
  final z = days + _b719468;
  final era = floorDiv(z, _b146097);
  final dayOfEra = (z - era * _b146097).toInt();
  final yearOfEra = (dayOfEra - dayOfEra ~/ 1460 + dayOfEra ~/ 36524 - dayOfEra ~/ 146096) ~/ 365;
  final dayOfYear = dayOfEra - (365 * yearOfEra + yearOfEra ~/ 4 - yearOfEra ~/ 100);
  final monthPrime = (5 * dayOfYear + 2) ~/ 153;
  final day = dayOfYear - (153 * monthPrime + 2) ~/ 5 + 1;
  final month = monthPrime + (monthPrime < 10 ? 3 : -9);
  final carry = month <= 2 ? BigInt.one : BigInt.zero;
  return (year: BigInt.from(yearOfEra) + era * _b400 + carry, month: month, day: day);
}

bool isLeapYear(BigInt year) =>
    floorMod(year, _b4) == BigInt.zero &&
    (floorMod(year, _b100) != BigInt.zero || floorMod(year, _b400) == BigInt.zero);

int daysInMonth(BigInt year, int month) =>
    month == 2 ? (isLeapYear(year) ? 29 : 28) : (const {4, 6, 9, 11}.contains(month) ? 30 : 31);

typedef Clock = DateTime Function();

/// The one sanctioned place where host wall-clock time enters domain math: the
/// only read of "now" anywhere in ChronoLog, converted once into an exact day
/// ordinal on the shared days axis. A caller needing a coarser ordinal
/// truncates this result and must not fork a second helper, because the "no
/// Date arithmetic in domain code" rule only holds while every caller shares
/// this conversion. The 24/1440/86400 below are the host operating system's
/// own civil units — the only units a [DateTime] can report — and are NOT the
/// document's coordinate law; how the resulting ordinal reads as hours is the
/// governing frame's law to decide.
Rational nowDays([Clock? clock]) {
  final now = (clock ?? DateTime.now)();
  final date = daysFromCivil(BigInt.from(now.year), now.month, now.day);
  return Rational(date) +
      Rational.fromInt(now.hour, 24) +
      Rational.fromInt(now.minute, 1440) +
      Rational.fromInt(now.second, 86400);
}
